ops/main.py

```diff
@@ -30,24 +30,43 @@ def _get_charm_dir():
 def _load_metadata(charm_dir):
     with open(charm_dir / 'metadata.yaml') as f:
         metadata = yaml.load(f, Loader=yaml.SafeLoader)
+
+    functions_meta = charm_dir / 'functions.yaml'
+    actions_meta = charm_dir / 'actions.yaml'
+    if functions_meta.exists():
+        # TODO: add a version check and a warning here to make sure functions are not skipped without notice pre-2.7.0 Juju versions.
+        if actions_meta.exists():
+            raise RuntimeError('charm must not mix functions and actions metadata')
+        metadata['functions_type'] = 'functions'
+        metadata['functions'] = yaml.safe_load(functions_meta.read_text())
+    elif actions_meta.exists():
+        metadata['functions_type'] = 'actions'
+        metadata['functions'] = yaml.safe_load(actions_meta.read_text())
     return metadata
 
 
-def _create_event_link(charm_dir, event_dir, target_path, bound_event):
+def _create_event_link(charm, bound_event):
     """Create a symlink for a particular event.
 
-    charm_dir -- A root directory of the charm
+    charm -- A charm object.
     bound_event -- An event for which to create a symlink.
     """
-    # TODO: Handle function/action events here.
-    if not issubclass(bound_event.event_type, ops.charm.HookEvent):
+    if issubclass(bound_event.event_type, ops.charm.HookEvent):
+        event_dir = charm.framework.charm_dir / 'hooks'
+        event_path = event_dir / bound_event.event_kind.replace('_', '-')
+    elif issubclass(bound_event.event_type, ops.charm.FunctionEvent):
+        event_dir = charm.framework.charm_dir / charm.framework.meta.functions_type
+        # The event_kind is suffixed with "_function" while the executable is not.
+        event_path = event_dir / bound_event.event_kind[:-len('_function')].replace('_', '-')
```

Comment:
We need something like this here, to avoid very weird bugs:
```
if not bound_event.event_kind.endswith("_function"):
        raise RuntimeError("function event name {event_name} needs _function suffix")
```

Reply:
Good point.